File: 프로그래머스/2/131130. 혼자 놀기의 달인/혼자 놀기의 달인.py
def solution(cards):
    answer = []
    
    for i in range(len(cards)):
        tmp = []
        
        while cards[i] not in tmp:
            tmp.append(cards[i])
            i = cards[i] - 1
            
        if sorted(tmp) not in answer:
            answer.append(sorted(tmp))
            
        answer = sorted(answer, key=lambda x: -len(x))

    if len(answer) == 1:
        return 0
    
    return len(answer[0]) * len(answer[1])

File: 프로그래머스/2/131130. 혼자 놀기의 달인/혼자 놀기의 달인.java
import java.util.*;

class Solution {
    
    boolean[] visited;
    int[] cards;

    public int solution(int[] cards) {
        int answer = 0;
        visited = new boolean[cards.length];
        this.cards = cards;
        Queue<Integer> pq = new PriorityQueue<>(Collections.reverseOrder());
        
        for (int i = 0; i < cards.length; i++) {
            if (visited[i]) continue;
            
            pq.add(dfs(i));
        }
        
        if (pq.size() > 1) {
            answer = 1;
            answer *= pq.poll();
            answer *= pq.poll();
        }
        
        return answer;
    }
    
    public int dfs(int startIdx) {
        if (visited[startIdx]) {
            return 0;
        }
        
        visited[startIdx] = true;
        
        return dfs(cards[startIdx] - 1) + 1;
    }
}